src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs

```diff
@@ -0,0 +1,485 @@
+            where TACList : IReadOnlyList<AnalyzerConfig>
+        {
+            // Since paths are compared as ordinal string comparisons, the least nested
+            // file will always be the first in sort order
+            if (!analyzerConfigs.IsSorted(DirectoryLengthComparer))
+            {
+                throw new ArgumentException(
+                    "Analyzer config files must be sorted from shortest to longest path",
+                    nameof(analyzerConfigs));
+            }
+
+            var diagnosticBuilder = ArrayBuilder<Diagnostic>.GetInstance();
+            var allTreeOptions = ArrayBuilder<TreeOptions>.GetInstance(sourcePaths.Count);
+            var allAnalyzerOptions = ArrayBuilder<AnalyzerOptions>.GetInstance(sourcePaths.Count);
+
+            var allRegexes = PooledDictionary<AnalyzerConfig, ImmutableArray<Regex>>.GetInstance();
+            foreach (var config in analyzerConfigs)
+            {
+                // Create an array of regexes with each entry corresponding to the same index
+                // in <see cref="EditorConfig.NamedSections"/>.
+                var builder = ArrayBuilder<Regex>.GetInstance(config.NamedSections.Length);
+                foreach (var section in config.NamedSections)
+                {
+                    string regex = AnalyzerConfig.TryCompileSectionNameToRegEx(section.Name);
+                    builder.Add(new Regex(regex, RegexOptions.Compiled));
+                }
+
+                Debug.Assert(builder.Count == config.NamedSections.Length);
+
+                allRegexes.Add(config, builder.ToImmutableAndFree());
+            }
+
+            Debug.Assert(allRegexes.Count == analyzerConfigs.Count);
+
+            var treeOptionsBuilder = ImmutableDictionary.CreateBuilder<string, ReportDiagnostic>(
+                CaseInsensitiveComparison.Comparer);
+            var analyzerOptionsBuilder = ImmutableDictionary.CreateBuilder<string, string>(
+                CaseInsensitiveComparison.Comparer);
+            foreach (var sourceFile in sourcePaths)
+            {
+                var normalizedPath = pathWithForwardSlashSeparators(sourceFile);
+                // The editorconfig paths are sorted from shortest to longest, so matches
+                // are resolved from most nested to least nested, where last setting wins
+                foreach (AnalyzerConfig config in analyzerConfigs)
+                {
+                    if (normalizedPath.StartsWith(config.NormalizedDirectory, StringComparison.Ordinal))
+                    {
+                        int dirLength = config.NormalizedDirectory.Length;
+                        // Leave '/' if the normalized directory ends with a '/'. This can happen if
+                        // we're in a root directory (e.g. '/' or 'Z:/'). The section matching
+                        // always expects that the relative path start with a '/'. 
+                        if (config.NormalizedDirectory[dirLength - 1] == '/')
```

Comment:
How do we know that NormalizedDirectory isn't an empty string?

Reply:
The code requires absolute paths be provided, and empty string is not a valid absolute path.

Reply:
I'll add argument validation for this.

Reply:
Turns out the AnalyzerConfig constructor will throw if this is not the case.